src/rt-app_taskgroups.c

```diff
@@ -373,7 +435,35 @@ static int cgroup_attach_task(char *name)
 	FILE *tasks;
 	size_t size;
 
-	file = strcmp(name, "/") ? "/tasks" : "tasks";
+	if (ctrl.version == 1) {
+		file = strcmp(name, "/") ? "/tasks" : "tasks";
+	}
+
+	if (ctrl.version == 2) {
+		if (strcmp(name, "/")) {
+			size_t path_size = strlen(ctrl.mount_point) + strlen(name) + strlen(TYPE) + 3;
+			char* cgroup_path = malloc(path_size);
+			FILE* cgroup_file;
+
+			snprintf(cgroup_path, path_size, "%s/%s/%s", ctrl.mount_point, name, TYPE);
+			cgroup_file = fopen(cgroup_path, "we");
+			if (!cgroup_file) {
+				perror("open");
+				goto error;
+			}	
+			if (fprintf(cgroup_file, "threaded") < 0) {
+				perror("fprintf");
+				goto error;
+			}
+			if (fclose(cgroup_file)) {
+				perror("fclose");
+				goto error;
+			}
+			free(cgroup_path);
+		}
+		file = strcmp(name, "/") ? "/cgroup.procs" : "cgroup.procs";
```

Comment:
Even when you switch the cgroup to `threaded` mode a write through `cgroup.procs` would migrate whole process (threadgroup) into target cgroup. The difference may not be visible with single-threaded processes. I assume rt-app is supposed to work with invididual threads of  a process.

OTOH, you may not always write only into `cgroup.threads`, e.g. because you can't pick a single thread of a (foreign) process and move it between cgroups that are not members of the same threaded subtree. (Not sure if this is the mode how rt-app operates.)

Reply:
rt-app is a single process that creates several threads so you should use cgroup.threads instead of cgroup.procs to move each thread in different cgroup as mentioned by @Werkov.